esp/services/ws_dfsservice/ws_dfsservice.hpp

```diff
@@ -0,0 +1,62 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2021 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#ifndef _WS_DFSSERVICE_HPP__
+#define _WS_DFSSERVICE_HPP__
+
+#include "ws_dfs_esp.ipp"
+
+#include "dadfs.hpp"
+#include <atomic>
+
+
+class CWsDfsEx : public CWsDfs
+{
+    std::atomic<bool> m_daliDetached{false};
+
+public:
+    IMPLEMENT_IINTERFACE;
+    virtual ~CWsDfsEx(){};
+    virtual void init(IPropertyTree *cfg, const char *process, const char *service);
+    virtual bool onGetLease(IEspContext &context, IEspLeaseRequest &req, IEspLeaseResponse &resp);
+    virtual bool onKeepAlive(IEspContext &context, IEspKeepAliveRequest &req, IEspKeepAliveResponse &resp);
+    virtual bool onDFSFileLookup(IEspContext &context, IEspDFSFileLookupRequest &req, IEspDFSFileLookupResponse &resp);
+
+private:
+
+    bool attachServiceToDali() override
```

Comment:
not likely to be thread safe (when it is eventually called).  You would want a test and clear/test and set.   I suspect you would need a critical section to actually code.

Reply:
I'm not sure these methods are needed at all... the service can't work without Dali..

I see there's a 'canDetachFromDali()' in e.g. CHttpSoapBinding that's overriden (to return false) in a few places, I suspect this service should be one of those cases, and then perhaps these attach/detach/isDetached method do not need implementing/overriding.

@wangkx @afishbeck - could you advise?

Reply:
When a detach from DALI is attempted, if any running service returns false from canDetachFromDali() the detach should fail.  There are several services which can't function without DALI.  There is a force flag but it would be used at the users own risk.  It does look like some DALI dependent services do keep track of calls detachServiceFromDali() just in case, and throw an exception if methods are called after detaching.  I think this is in case the force option is used with detach.

```
bool CWsDfuEx::DFUDeleteFiles(IEspContext &context, IEspDFUArrayActionRequest &req, IEspDFUArrayActionResponse &resp)
{
    if (isDetachedFromDali())
        throw MakeStringException(ECLWATCH_INVALID_INPUT, "ESP server is detached from Dali. Please try later.");
```
If in the future you are doing things like maintaining file locks, etc.  then we could look at not even allowing forced detach. I don't know that forced detach has much value, it's a bad configuration / combination of services if it's needed.
 Perhaps not allowing it if you throw an exception when force were used.  
@rpastrana would be the best to ask the details

Reply:
I think the setup/conifguration should be such that the esp for this service, is never requested to detach from Dali..
But is it always via a client request (DetachFromDaliRequest).
Are there circumstances where esp notices Dali is detected (down)..

e.g. if esp can handle or in future handle a Dali restarting, and notifies services via this same mechanism, then it may make sense for this/other services to implement these methods as a flag and throw..
@rpastrana - could you comment?

Reply:
For now I think it's okay to leave it with defaults, the worst that can happen is that if a dali detach happens and effects this service a SDS exception will ensure.

Reply:
Here's the original PR for reference: https://github.com/hpcc-systems/HPCC-Platform/pull/10832 
In short, the detach from dali feature was meant to be a safeguard non ECLWatch services from being affected by unresponsive Dali.

The forced detach requirement seems to have come from "historical fears of [transactional ESP services] being attached to DALI in production".

Now that transactional ESP services are further detached from ECLWatch/ECLServices in containerized mode, perhaps this feature should be #ifdefed away (in non-baremeta)?

For now I'd decline to detach if the service cannot function without Dali.